cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -0,0 +1,2987 @@
+use regalloc::{Reg, RegClass};
+
+use crate::isa::x64::inst::*;
+
+fn low8willSXto64(x: u32) -> bool {
+    let xs = (x as i32) as i64;
+    xs == ((xs << 56) >> 56)
+}
+
+fn low8willSXto32(x: u32) -> bool {
+    let xs = x as i32;
+    xs == ((xs << 24) >> 24)
+}
+
+//=============================================================================
+// Instructions and subcomponents: emission
+
+// For all of the routines that take both a memory-or-reg operand (sometimes
+// called "E" in the Intel documentation) and a reg-only operand ("G" in
+// Intelese), the order is always G first, then E.
+//
+// "enc" in the following means "hardware register encoding number".
+
+#[inline(always)]
+fn mkModRegRM(m0d: u8, encRegG: u8, rmE: u8) -> u8 {
+    debug_assert!(m0d < 4);
+    debug_assert!(encRegG < 8);
+    debug_assert!(rmE < 8);
+    ((m0d & 3) << 6) | ((encRegG & 7) << 3) | (rmE & 7)
+}
+
+#[inline(always)]
+fn mkSIB(shift: u8, encIndex: u8, encBase: u8) -> u8 {
+    debug_assert!(shift < 4);
+    debug_assert!(encIndex < 8);
+    debug_assert!(encBase < 8);
+    ((shift & 3) << 6) | ((encIndex & 7) << 3) | (encBase & 7)
+}
+
+#[inline(always)]
+// Get the encoding number from something which we sincerely hope is a real
+// register of class I64.
+fn iregEnc(reg: Reg) -> u8 {
+    debug_assert!(reg.is_real());
+    debug_assert!(reg.get_class() == RegClass::I64);
+    reg.get_hw_encoding()
+}
+
+// F_*: these flags describe special handling of the insn to be generated.  Be
+// careful with these.  It is easy to create nonsensical combinations.
+const F_NONE: u32 = 0;
+
+// Emit the REX prefix byte even if it appears to be redundant (== 0x40).
+const F_RETAIN_REDUNDANT_REX: u32 = 1;
+
+// Set the W bit in the REX prefix to zero.  By default it will be set to 1,
+// indicating a 64-bit operation.
+const F_CLEAR_REX_W: u32 = 2;
+
+// Add an 0x66 (operand-size override) prefix.  This is necessary to indicate
+// a 16-bit operation.  Normally this will be used together with F_CLEAR_REX_W.
```

Comment:
```suggestion
/// Add an 0x66 (operand-size override) prefix.  This is necessary to indicate
/// a 16-bit operation.  Normally this will be used together with F_CLEAR_REX_W.
```